Fix bugs in sqlite timeseries output and add IDF parsing capabilities within unit tests

mbadams5 opened the pull request:
This fixes two bugs in the SQLite time series data where interval periods are incorrect or meters are missing time indexes. I added a bunch of unit tests to test upstream of the SQLite class. There are now unit tests at each of the points in OutputProcessor where it interacts with the SQLite class. 

This should fix both bugs from my unit tests and manual testing, although I would like other people to test this branch as well. 

I also found a bug in DetermineIndexGroupKeyFromMeterName from adding unit tests. Also, DateToStringWithMonth didn't have enough validation to pass all my new unit tests so I refactored the code after creating quite a few failing tests. That function probably never got passed invalid data so it was never an issue but now it is more robust.

I also moved the SQLiteFixture to its own header file in a 'fixtures' folder in the unit test folder. I did this since it is now used by both the SQLite.unit.cc and OutputProcessor.unit.cc

===================== Update =========================

Well this one definitely got away from me quickly.... What started out as a fairly simple bug fix turned into massive Input/Output testing and a new unit testing framework to process IDFs. This is the huge new addition. Now you can pass in an IDF snippet and have it run through the normal InputProcessor routines to fill out the global data structures. This paves the way to call GetFooInput() for any object without having to generate that global data yourself. Generating the data is error prone and can easily cause issues between different unit tests as @Nigusse and others have noticed.

So now in a unit test you use the EnergyPlusFixture (or any derived fixtures from that fixture) and this allows you to call process_idf(). You need to pass in a valid IDF snippet otherwise process_idf() will return false and may output more specific error messages as well (the same as ProcessInput() in InputProcessor). Another advantage of these fixtures is that they clear the global state of the various namespaces on TearDown() meaning at the end of every individual unit test run. This is done through a new clear_state() function in the namespace (not all namespaces have it yet, I was going through on an as need basis for now). As part of all this, the EnergyPlusFixture reads in the full Energy+.idd (that is output to the Products folder) and after the first call, will cache it. Then all subsequent calls (within that same executable run) should use this cached IDD saving time on unit testing. 

I refactored some of the Humidifiers, MixedAir, and HeatBalanceManager unit tests to take advantage of this new framework and to show how to use it. There are also examples in the many new unit tests for OutputProcessor and InputProcessor as well as unit tests for my new helper classes.

The pattern i've started is that all fixtures derive from EnergyPlusFixture and you only create a new fixture if it adds needed functionality not captured in EnergyPlusFixture AND is not appropriate to add to EnergyPlusFixture. I tried to make a hierarchy and start making the fixtures adhere to OOP.

I think we should separate DeathTests from normal tests. An example of a death test is [here](https://github.com/NREL/EnergyPlus/blob/460124da779f68972a3cb2f8eeba501e6f399acd/tst/EnergyPlus/unit/OutputProcessor.unit.cc#L2146). These are tests where you expect the program to not exit normally. In this case, ShowFatalError() is called and exits EnergyPlus. This allows us to test code paths that will cause program termination and make sure it happens how we expect.

You can now compare an expected string against the ESO, MTR, ECHO, COUT, and CERR streams as well as see if there is or is not any output on those streams.

There is a compare_containers function that will compare two different containers (either STL OR ObjexxFCL). It will make sure the two containers are the same size then iterate through each item in the container comparing the items. If they differ it will show the difference and at what index (zero based) they differ. Example usage...

``` c++
EXPECT_TRUE( compare_containers( std::vector< bool >( { true } ), ObjectDef( index ).AlphaOrNumeric ) );
EXPECT_TRUE( compare_containers( Array1D_bool( { true } ), ObjectDef( index ).AlphaOrNumeric ) );
```

Oh and just for you @mjwitte, the EnergyPlusFixture will automatically call ShowMessage( "Begin Test: " ) on every individual unit test run, putting the test case name and test name into that message. No need to put that message into every unit test anymore.

The overall unit test coverage increased from 5.49% to 6.99%. InputProcessor.cc went from 7.0% to 58.9% and OutputProcessor.cc went from 17.3% to 65.6%.

@kbenne @macumber @asparke2 @Myoldmopar 

Fixes #4979 
[Fixes #97920946]

Fixes #4982 
[Fixes #97921020]

[Delivers #98501064]

mbadams5 committed:
Fix sqlite bugs and add more unit tests upstream

This fixes #4979 and fixes #4982

mbadams5 committed:
Fix unit test failures

mbadams5 committed:
Merge branch 'develop' into sqlite_timeseries

mbadams5 committed:
Fix windows newline issue (for real hopefully)

mbadams5 committed:
Add more unit tests

Still might need to add more…

mbadams5 committed:
Merge branch 'develop' into sqlite_timeseries

mbadams5 committed:
Add more unit test

Found a bug while adding new unit test

mbadams5 committed:
Fix minor bug not related to sqlite

The issue is that ‘InteriorLights:Electricity’ is a substring of
‘InteriorLights:Electricity:Zone’ thus it is always stopping at the
former even when the latter is requested. Rearranged the order so this
does not happen. It passes the unit test now.

mbadams5 committed:
Add more unit tests

Found a bug in DateToStringWithMonth although it may just always get
passed validate data….

mbadams5 committed:
Add extra validation to DateToStringWithMonth

When adding unit tests to this function I found it lacked validation
around edge cases. I refactored this function and now it passes all my
unit tests. This function probably never gets fed invalidate data
during an actual simulation but this makes the function more robust
anyways.

mbadams5 committed:
Clean up to get tests to appear in CI

Apparently the regex on the CI to find unit test macro doesn’t like a
space between TEST or TEST_F and the parentheses. I guess????

mbadams5 committed:
Add more unit tests and clean up tests

mbadams5 commented:
I may or may not add more unit tests...

mbadams5 committed:
Add more unit tests

mbadams5 commented:
@DeadParrot 

On Linux/Mac, OutputProcessor::WriteRealData outputs 1.0e-2 as "1.000000000000000E-02" but on Windows it seems to output "1.000000000000000E-002". I can't seem to find a reason for this apparent platform difference. This shows up in the single failing Windows unit test. Thoughts?

DeadParrot commented:
@mbadams5 This is a known platform difference. You can [override it](https://msdn.microsoft.com/en-us/library/0fatw238.aspx) with `_set_output_format(_TWO_DIGIT_EXPONENT);` in an MSVC `#ifdef` but we'd have to make sure it can still grow to 3 digits when needed. I think Intel C++ follows MS on this but I'd have to check on current MinGW behavior. If you don't have all these compilers and want me to check into the exact behavior let me know.

**Edit**: Had a chance to try this out and I found that the same call is needed and works for VC, Intel, and MinGW GCC on Windows if you want 2-digit exponents when 3 aren't needed. Just putting this at the top of `main` seems to do the trick:

``` cpp
#ifdef _WIN32
    _set_output_format(_TWO_DIGIT_EXPONENT);
#endif
```

There's a chance some other/future Windows compiler won't support this but it is probably logical to assume most do/will.

This change could break other tests and/or alter outputs.

mbadams5 committed:
Try to fix failing Windows unit test

mbadams5 commented:
Humm... well I tried a different IDF and while it still has the fixes for these bugs, each of the records in the Time table are doubled. I think this is what I originally attempted to solve. I will write new unit tests to expose this behavior and attempt to fix the doubling of time entries.

mbadams5 committed:
Work on more unit tests

mbadams5 committed:
Merge branch 'develop' into sqlite_timeseries

mbadams5 committed:
Add EnergyPlusFixture and OutputProcessorFixture

mbadams5 committed:
Clean up EnergyPlusFixture

mbadams5 committed:
Add documentation, add compare cerr/cout

mbadams5 committed:
Get rid of cerr output in DXCoil unit test

mbadams5 committed:
Add InputProcessor unit tests

This will currently read in an IDD snippet or the Full Energy+.idd file
in the Products folder and process it. The fixture takes care of all
deallocations in the TearDown() phase so no extra steps are needed in
the unit tests themselves.

This still needs some work. The InputProcessorFixture should probably
be in its own header file OR this should go in the EnergyPlusFixture
since this usage will probably be common between a lot of different
unit tests.

mbadams5 committed:
Try to fix failing unit tests and build failures

mbadams5 committed:
Try to fix failing unit test again

This should hopefully work on the CI machines to find the Energy+.idd
file…

mbadams5 committed:
Fix failing unit tests and add IDF parsing

mbadams5 committed:
Major clean up and fix bad unit tests

mbadams5 committed:
Add new unit test

mbadams5 committed:
Implement IDD cache, show message, clean up

This implements IDD caching since it is now part of EnergyPlusFixture
SetUpTestCase(). These means it will be run once when the Fixture is
setup before any test cases are run based on that fixture. I did this
because I eliminated the option to pass an IDD snippet to processIDF
and made processIDD private. I was worried that the IDD snippets would
be incorrect or become out of date as Energy+.idd is updated. So the
full IDD is now read unless the cached IDD is used.

One thing is that the IDD is parsed for caching for any Fixture that
inherits from EnergyPlusFixture as well. This happens by default IF
static void SetUpTestCase() {} is not part of the derived class. If
that code is in the derived class then it will not initialize the
m_idd_cache. This is important to know because if you need to parse an
IDF then you need to add EnergyPlusFixture::SetUpTestCase(); to that
function like so: static void SetUpTestCase() {
EnergyPlusFixture::SetUpTestCase(); }

I also typedef’d the fixtures as FooDeathTestFixture so that we can
have separate death test only unit tests but they retain everything
about the fixture.

I also created a show_message() function that grabs the
test_case_name() and name() from the unit test and creates the
ShowMessage text that was in each unit test. This gets called every
time EnergyPlusFixture or any derived fixture is used. This eliminates
the need to put this text in every unit test as well as automatically
updates every time the unit test name changes.

mbadams5 committed:
Refactor names and remove unnecessary nullptr

mbadams5 committed:
Add more unit tests, clean up code

mbadams5 committed:
Move ReportNumberCounter, fix build error

Apparently clang has an extension to allow construction of tuple by
initialization list the way I was doing it but the other compilers
complained. Now I am using make_tuple instead.

Moved ReportNumberCounter out from static variable in
AssignReportNumber() into OutputProcessor namespace. This was done to
be able to reset the counter for unit testing. It was causing
inconsistent unit test results depending on run order.

mbadams5 committed:
Add more unit tests and failing unit test for bug fix

mbadams5 committed:
FINALLY fix original bugs

This finally fixes the original bugs and gets rid of the double
timestamps in the Time table for the SQL output.

mbadams5 committed:
Clean up unit test

mbadams5 committed:
Fix unit test and linux warning

This pulls four static variables out of UpdateDataandReport and places
them in the namespace instead. This shouldn’t change anything for a
normal simulation, but allows for unit testing to reset those variables.

mbadams5 committed:
This is a major clean up and adds a custom parser

The parser is needed for the IDF snippets since it now looks for all
required objects in a snippet. So instead of making everyone add
Building and GlobalGeometryRules, it now looks for those objects and
adds them if they are not in the IDF. Otherwise it uses the existing
objects. The parser is a recursive descent parser so it is quick.

I started to add documentation to the new functions. There needs to be
more still.

I also still need to fix a few lingering issues that will help
developers to use this more easily.

There are also now unit tests for the EnergyPlusFixture under
EnergyPlusMetaFixture (so meta…)

mbadams5 committed:
Add clear_state(), move function static, clean up

First, this adds a clear_state() function to various namespaces. This
function will do as it says, clear the state of the namespace. This is
important for unit testing. This function should never be used during
normal simulation. This functionality is basically moved from
EnergyPlusFixture, but having it in the namespace allows for the next
change.

Moved various InputProcessor function static variables into the
namespace. However, for both InputProcess and OutputProcessor these
variables are not in the header file as ‘extern foo’. This means no one
outside of that implementation file can use those variables. This
maintains some of the original intend. However, this means those
variables can now be reset via the clear_state() function.

mbadams5 committed:
Merge branch 'develop' into sqlite_timeseries

mbadams5 committed:
Add unit tests, fix unit failures, clean up code

mbadams5 committed:
Documentation and clean up

mbadams5 committed:
Merge branch 'develop' into sqlite_timeseries

mbadams5 committed:
Major clean up and more documentation

I cleaned up my EnergyPlusFixture. I pulled the embedded classes out
into their own files. This allowed for better (and cleaner) testing.

macumber commented:
@mbadams5 I did some light testing and both issues I had seen appear to be fixed

mbadams5 committed:
Add more clear_state() and refactor Humidifiers tests

The Humidifiers unit tests are the same as before functionality-wise
however I removed all the code needed to set up state related to input.
Instead it now uses the new process_idf() function to set up that state.

mbadams5 committed:
Refactor HeatBalanceManager tests

mbadams5 committed:
Add more clear_state() and clean up

mbadams5 committed:
Refactor MixedAir unit test

mbadams5 committed:
Fix compile errors

mbadams5 committed:
Fix failing windows unit test

mbadams5 committed:
Change idf checks and clean up OutputProcessor test

mbadams5 commented:
Check out the updated PR text. Lots of new changes. Would like some eye balls on this.

@Myoldmopar @kbenne @mjwitte @rraustad @Nigusse

mbadams5 commented on tst/EnergyPlus/unit/Humidifiers.unit.cc:
```diff
@@ -4,411 +4,245 @@
+
+		ASSERT_FALSE( process_idf( idf_objects ) );
+
+		GetHumidifierInput();
```
@Nigusse I changed this test to use the new input processing capabilities. The assertions are the same as you had  and the only thing I changed was remove the IDD and IDF data structures creation. Instead it now uses process_idf() and the IDF snippet. Make sure this test is still what you intended, but I think it is.

mbadams5 commented on tst/EnergyPlus/unit/Humidifiers.unit.cc:
```diff
@@ -4,411 +4,245 @@
+		thisHum.EfficiencyCurvePtr = CurveManager::GetCurveIndex( "THERMALEFFICIENCYFPLR" );
+		thisHum.EfficiencyCurveType = CurveManager::Quadratic;
+
+		thisHum.CalcGasSteamHumidifier( 0.030 );
```
@Nigusse I changed this test to use the new input processing capabilities. The assertions are the same as you had  and the only thing I changed was remove the IDD and IDF data structures creation. Instead it now uses process_idf() and the IDF snippet. Make sure this test is still what you intended, but I think it is.

mbadams5 commented on tst/EnergyPlus/unit/MixedAir.unit.cc:
```diff
@@ -4,153 +4,100 @@
+		ErrorsFound = false;
+		ProcessOAControllerInputs( CurrentModuleObject, ControllerNum, AlphArray, NumAlphas, NumArray, NumNums, lNumericBlanks, lAlphaBlanks, cAlphaFields, cNumericFields, ErrorsFound );
+		EXPECT_FALSE( ErrorsFound );
+		EXPECT_EQ( 6, OAController( 2 ).OANode );
```
@mjwitte  I changed this test to use the new input processing capabilities. All the assertions, except one, are the same as you had  and the only thing I changed was remove the IDD and IDF data structures creation. Instead it now uses process_idf(), the IDF snippet, and a couple other normal E+ functions to setup data structures. I had to change this line from 

``` c++
EXPECT_EQ( 4, OAController( 2 ).OANode );
```

to

``` c++
EXPECT_EQ( 6, OAController( 2 ).OANode );
```

since this now uses the fully defined IDF object and it counts the Relief and Return Air Nodes of the first Controller:OutdoorAir where it did not define those Nodes in your previous code. Make sure this test is still what you intended, but I think it is.

rraustad replied:
This looks correct.

mjwitte replied:
@mbadams5 @rraustad On the surface that looks fine.  I would have to go back and find the original pull request when this test was added to see the original problem that was being tested.  There may have been a reason for leaving some of the nodes blank.  Lesson learned - unit tests should have comments documenting why it was set up the way it was.

mbadams5 replied:
@mjwitte #4715 was the PR and #4272 was the issue. Reading through it, it seems like you were just testing to allow an actuator node name that was not an OutdoorAir:Node. But you might have more clear insight. 

That is a good point about documenting unit tests, especially for bug fixes.

mjwitte replied:
@mbadams5 Thanks for finding the references.  This looks ok then.

rraustad commented on tst/EnergyPlus/unit/Fixtures/EnergyPlusFixture.hh:
```diff
@@ -0,0 +1,194 @@
+		// This function processes an idf snippet and defaults to using the idd cache for the fixture.
+		// The cache should be used for nearly all calls to this function.
+		// This more or less replicates InputProcessor::ProcessInput() but in a more usable fashion for unit testing
+		bool process_idf( std::string const & idf_snippet, bool use_idd_cache = true ) {
```
So process_idf DOES NOT test the input processor? Will ProcessInput eventually be replaced with this function?

mbadams5 replied:
process_idf does not test InputProcessor. It uses the same functions in InputProcessor however. This is really just a helper function for unit testing only. I test (most) of InputProcessor in InputProcessor.unit.cc. This new process_idf() function wraps all needed InputProcessor functions in a much more unit testing friendly way, instead of just calling InputProcessor::ProcessInput(). This new function also allows for passing only IDF snippets and more importantly, passing them as strings instead of opening IDF files.

Myoldmopar commented:
@mbadams5 This is enormous, and very cool to see idf snippets working so well.  From an E+ standpoint, the changes were very minor, as long as you don't randomly call clear_state().  Also from an E+ standpoint, there are no diffs --- at all.  So that is wonderful.

For the original root of this pull request, I am not that familiar with the sqlite issues, but I was following this in the earlier commits as you were adding unit tests to check the output, and that those now pass.  Maybe as a wrap up of this PR, and how deeply you have been with the unit tests, you could provide some guidance on whether we need (and if so, how to integrate) further SQLite testing, or your suggestions for enhancements to the unit testing framework.

I am happy to merge this, there is no reason for it to linger.  We can update the team on Wednesday and maybe show off the new idf snippet based unit testing quite easily.  One thing to keep in mind though.  This will allow developers to call GetInput, which will provide an initial ease on unit testing some namespaces, because you can then call the other worker functions within the namespace with an index to the main array.  This propagates an array-index-based EnergyPlus, which isn't the eventual path we want to be on, so guidance should definitely include both short-term and long-term unit test work.  

@kbenne @mjwitte @mbadams5 I'll wait a bit today in case you have anything further, but if I don't hear anything, I'll merge in later today.

kbenne commented:
This is a lot to give a reliable review.  I'l start with some responses to @mbadams5's comment at the top of the PR.  I wish I could line comment on comments like I can do in code.

"As part of all this, the EnergyPlusFixture reads in the full Energy+.idd (that is output to the Products folder) and after the first call, will cache it. Then all subsequent calls (within that same executable run) should use this cached IDD saving time on unit testing."

Just note that this has no value to the CI because it uses ctest which starts every test in a brand new process.

I'm not a fan of type defing the test fixture.  Why not just call a spade a spade and use that as the test name?  If your humidifier tests use the EnergyPlusFixture or the HVACFixture or whatever, why not just use that in the test macro for your name?  You're only serving to obfuscate things, however minimally.  - bike shed - because there is way too much advanced goodness going on in here for me to catch and comment on it all.

"Oh and just for you @mjwitte, the EnergyPlusFixture will automatically call ShowMessage( "Begin Test: " ) on every individual unit test run, putting the test case name and test name into that message. No need to put that message into every unit test anymore."

Isn't the test runner doing this for you anyway.  This seems like unwanted noise.  I'm sure @mjwitte has a good reason that I am not grasping, but maybe there are features in the test runner that people are not aware of.  Like verbose output.

The ability to load idf snippets seems cool and much needed.  But I understand this was implemented separately from the idf parser that EnergyPlus main routine normally uses.  Hey you called this out on the phone to me @mbadams5.  Why'd you do that, now I just have something to complain about.  Is it technically challenging to extract this part out of energyplus so everyone can use it?

kbenne commented on src/EnergyPlus/DataEnvironment.cc:
```diff
@@ -181,6 +181,122 @@ namespace DataEnvironment {
+	// Clears the global data in DataEnvironment.
+	// Needed for unit tests, should not be normally called.
+	void
+	clear_state()
```
I'm curious what is the analog if this if the namespaces were evolved into classes?  Would you just destroy the object and recreate a new one.  Or would you still have this method exposed publicly?

kbenne replied:
You seem to be setting up a scenario where namespaces could become classes and perhaps their ctor take idf text or alternatively a data structure produced by your parser.  ??

kbenne replied:
I'm in favor of that view of the world.  Data gets classified as time vary or not time varying.  Class are instantiated with all not time varying data, possibly from idf parser.  time varying data must be passed into and out of member functions.

mbadams5 replied:
I think generally most of the namespace variables become member variables of a class and, yes, get recreated on any new object. There may be cases where this needs to be publicly exposed still, but I can't think of any off the top of my head.

Yes it would be nice if each class got instantiated by a data structure parsed by the InputProcessor. I'll address your comment about 'my' parser in a different comment.

I tend to agree with your 3rd comment.

mbadams5 commented:
@kbenne I agree, inline commenting would be nice...

> Just note that this has no value to the CI because it uses ctest which starts every test in a brand new process.

This is a general topic that needs to be covered with the team. How ctest (and thus the CI) could have subtle differences compared to straight Google Tests (aka energyplus_tests). Also, I reworked the caching so it is lazy instantiated now so it is not used EVERY fixture load. This helped even the ctest version since not all tests need to use the cached IDD.

> I'm not a fan of type defing the test fixture. Why not just call a spade a spade and use that as the test name? If your humidifier tests use the EnergyPlusFixture or the HVACFixture or whatever, why not just use that in the test macro for your name? You're only serving to obfuscate things, however minimally. - bike shed - because there is way too much advanced goodness going on in here for me to catch and comment on it all.

If there is general consensus to not have the typedef'd names then i'm fine with that. I think it helps to avoid name collisions and make things more clear when you run ./energyplus_tests or ctest. But i'm ok with just calling a "spade a spade" if every wants that instead.

> Isn't the test runner doing this for you anyway. This seems like unwanted noise. I'm sure @mjwitte has a good reason that I am not grasping, but maybe there are features in the test runner that people are not aware of. Like verbose output.

Kyle you are missing why @mjwitte did ShowMessage( "Begin Test: " ) to begin with. That will output the text to the eplusout.err file. I think he did that so he could look at the error file and know which lines of error file text are for which specific unit test. The error file is not an easy file (right now) to close and delete on every unit test run. So this is an intermediate step in my mind. It gets rid of all the ShowMessage() in every individual unit tests but keeps that message until the error file can be deleted every unit test.

> The ability to load idf snippets seems cool and much needed. But I understand this was implemented separately from the idf parser that EnergyPlus main routine normally uses. Hey you called this out on the phone to me @mbadams5. Why'd you do that, now I just have something to complain about. Is it technically challenging to extract this part out of energyplus so everyone can use it?

This was NOT implemented separately from the idf parser that EnergyPlus main routine normally uses. Again, NOT separate. This uses the E+ functions to parse the IDF snippet and fill in the global data structures.  That said, I did implement my own (naive) idf parser to pre-parse the IDF snippet before running it through the normal E+ idf parser. This was done because E+ has required-objects that must be in the IDF for it to be valid. These objects (as of now) are Building and GlobalGeometryRules. So instead of making everyone input those objects into their IDF snippets, I search their snippet for those objects. If the objects are there then I use them, and if not, then I put in a default object so the snippet is a valid IDF. That is all I use my custom IDF parser for and it is only for use within unit tests and the EnergyPlusFixture. As an aside, I originally tried to do this using a regular expression. HOWEVER, gcc has a (VERY) incomplete <regex> implementation until 4.9.0 so no matter how I tried to structure my regex it wouldn't work.

kbenne commented:
so the idf parser was separate.. :) 

seriously pardon my confusion.

kbenne commented on tst/EnergyPlus/unit/Fixtures/EnergyPlusFixture.hh:
```diff
@@ -0,0 +1,194 @@
+		// 		EXPECT_TRUE( compare_containers< Array1D_bool >( { true } , ObjectDef( index ).AlphaOrNumeric ) );
+		template < typename T, class T2 >
+		bool compare_containers( T const & expected_container, T2 const & actual_container ) {
+			return ioHelper->compare_containers< T, T2 >( expected_container, actual_container );
```
Why is there this level of indirection that seemingly does nothing but forward the call to IOHelper?  Is IOHelper inaccessible (ie private) or something?

Is this specifically for comparing an FCL type against a std type?  I assume if you just wanted to compare two std types you would use std::equal?  I bet you are using std::equal somewhere in the implementation. 

This seems like something universally useful beyond testing or io.  Maybe something for FCL directly?

mbadams5 replied:
Yes ioHelper is private member variable. I pulled IOHelper into its own class to be able to unit test that new functionality easier. I originally had it all within EnergyPlusFixture but it was an ugly mess and not easily testable.

It can compare FCL to FCL, STL to FCL, and STL to STL. I could use std::equal to do this, yes, but I wanted to be able to check size of containers as well as output the values within the containers that differ and at what index.

There could definitely be something within E+ or FCL itself that just uses std::equal to compare any container types.

kbenne commented on tst/EnergyPlus/unit/Fixtures/EnergyPlusFixture.hh:
```diff
@@ -0,0 +1,194 @@
+		// This function creates a string based on a vector of string inputs that is delimited by DataStringGlobals::NL by default, but any 
+		// delimiter can be passed in to this funciton. This allows for cross platform output string comparisons.
+		std::string delimited_string( std::vector<std::string> const & strings, std::string const & delimiter = DataStringGlobals::NL ) {
+			return ioHelper->delimited_string( strings, delimiter );
```
Same comment for all of the fowarding calls.  I'm sure you have a reason.  Please enlighten.

kbenne replied:
Almost looks like PIMPL.

kbenne commented on tst/EnergyPlus/unit/Fixtures/IdfParserFixture.hh:
```diff
@@ -0,0 +1,83 @@
+
+		void eat_whitespace( std::string const & idf, size_t & index ) {
+			IdfParser parser;
+			parser.eat_whitespace( idf, index );
```
How about a static method so you can avoid the cost and annoyance of instantiating parser?  Or a free function for that matter.  They're not evil.

mbadams5 replied:
I think I originally had it as static methods. I don't know why I changed it. Free functions could work as well. Let me think about it some.

kbenne commented on tst/EnergyPlus/unit/IdfParser.unit.cc:
```diff
@@ -0,0 +1,282 @@
+
+namespace EnergyPlus {
+
+	TEST_F( IdfParserFixture, decode ) {
```
Yep already disliking the type defs. Is this really EnergyPlusFixture?

mbadams5 replied:
This isn't even typedef'd. It is a true Fixture as seen above in...

``` c++
#include "Fixtures/IdfParserFixture.hh"
```

kbenne replied:
Yeah but now that I know the type defs are floating around I am always second guessing.  How do I know what you did in IdfParserFixture.hh?  You could have but a type def in there.  Yes once you learn the convention it might be fine, but people aren't going to implement the convention consistently.

kbenne commented on tst/EnergyPlus/unit/TestHelpers/IOHelper.cc:
```diff
@@ -0,0 +1,508 @@
+
+	bool IOHelper::compare_eso_stream( std::string const & expected_string, bool reset_stream ) {
+		auto const stream_str = this->eso_stream->str();
+		EXPECT_EQ( expected_string, stream_str );
```
This is surprising to me.  An assertion down in a supporting function.  Then you return the result of comparison.  Am I as a client of this thing supposed to put an assertion around that return value, thus doing the job twice?  Or I am I supposed to know you hide an assertion in your implementation?

kbenne commented on tst/EnergyPlus/unit/TestHelpers/IOHelper.hh:
```diff
@@ -0,0 +1,232 @@
+		// there are more macro arguments than expected. This is a way around that restriction.
+		// Example Usage: 
+		// 		EXPECT_TRUE( compare_containers< std::vector< bool > >( { true } , ObjectDef( index ).AlphaOrNumeric.begin() ) );
+		// 		EXPECT_TRUE( compare_containers< Array1D_bool >( { true } , ObjectDef( index ).AlphaOrNumeric.begin() ) );
```
How about no explicit template arguments?  Can't the compiler deduce if you give it a little help.  

`compare_containers(std::vector<bool>( { true } ), ObjectDef( index ).AlphaOrNumeric.begin() )`

I admit it is almost the same amount of typing, but it seems like it is a little better form.  And more leaves us less puzzled as to why one template argument is explicit and the other is not.

mbadams5 replied:
Yes the compiler could probably deduce if you passed like this. But just using `{ true }`` it could not.

kbenne commented:
One more thing and this is really the only thing that actually matters in all of my comments.  @mbadams5 and @macumber, now that Mark has fixed the database bug, do we need to take action in OS to remove the workaround for this particular version of E+?  Presumably that doesn't happen until we go through the OS update process, but it is important not to forget!

macumber commented:
The workaround here https://github.com/NREL/OpenStudio/blob/develop/openstudiocore/src/utilities/sql/SqlFile_Impl.cpp#L2396 should only get called for E+ 8.3, when we update to 8.4 it should no longer be triggered

mbadams5 committed:
Update and clean up based on @kbenne comments

mbadams5 committed:
Fix compare_containers usage to match documentation

mbadams5 committed:
Merge branch 'develop' into sqlite_timeseries

Myoldmopar commented:
AAAAHHHH!!!! @mbadams5 what just happened?
